src/discriminators/DiscriminatorTypeComposer.js

```diff
@@ -0,0 +1,338 @@
+      throw Error('Discriminator Key not Set, Use composeWithMongoose for Normal Collections');
+    }
+
+    opts = {  // eslint-disable-line
+      reorderFields: true,
+      customizationOptions: {
+        schemaComposer,
+      },
+      ...opts,
+    };
+
+    const baseTC = composeWithMongoose(baseModel, opts.customizationOptions);
+
+    const _DiscriminatorTypeComposer = this._getClassConnectedWithSchemaComposer(
+      opts.customizationOptions && opts.customizationOptions.schemaComposer
+    );
+    const baseDTC = new _DiscriminatorTypeComposer(baseTC.getType());
+
+    baseDTC.opts = opts;
+    baseDTC.childTCs = [];
+    baseDTC.discriminatorKey = (baseModel: any).schema.get('discriminatorKey') || '__t';
+
+    // discriminators an object containing all discriminators with key being DNames
+    baseDTC.DKeyETC = createAndSetDKeyETC(baseDTC, (baseModel: any).discriminators);
+
+    reorderFields(baseDTC, (baseDTC.opts: any).reorderFields, baseDTC.discriminatorKey);
+
+    baseDTC.DInterface = baseDTC._createDInterface(baseDTC);
+    baseDTC.setInterfaces([baseDTC.DInterface]);
+
+    baseDTC.schemaComposer.addSchemaMustHaveType(baseDTC);
+
+    // prepare Base Resolvers
+    prepareBaseResolvers(baseDTC);
+
+    return baseDTC;
+  }
+
+  _createDInterface(baseTC: DiscriminatorTypeComposer<any>): InterfaceTypeComposerClass<TContext> {
+    return this.schemaComposer.InterfaceTypeComposer.create({
+      name: `${baseTC.getTypeName()}Interface`,
+
+      resolveType: (value: any) => {
+        const childDName = value[baseTC.getDKey()];
+
+        if (childDName) {
+          return baseTC.schemaComposer.getTC(childDName).getType();
+        }
+
+        // as fallback return BaseModelTC
+        return baseTC.schemaComposer.getTC(baseTC.getTypeName()).getType();
+      },
+      fields: (): ComposeFieldConfigMap<any, TContext> => {
+        const baseFields = baseTC.getFieldNames();
+        const interfaceFields = {};
+        for (const field of baseFields) {
+          interfaceFields[field] = baseTC.getFieldConfig(field);
+        }
+        return interfaceFields;
+      },
+    });
+  }
+
+  getDKey(): string {
+    return this.discriminatorKey;
+  }
+
+  getDKeyETC(): EnumTypeComposer {
+    return this.DKeyETC;
+  }
+
+  getDInterface(): InterfaceTypeComposerClass<TContext> {
+    return this.DInterface;
+  }
+
+  hasChildTC(DName: string): boolean {
+    return !!this.childTCs.find(ch => ch.getTypeName() === DName);
+  }
+
+  setFields(fields: ComposeFieldConfigMap<any, any>): DiscriminatorTypeComposer<TContext> {
+    super.setFields(fields);
+
+    for (const childTC of this.childTCs) {
+      childTC.setFields(fields);
+    }
+
+    return this;
+  }
+
+  setField(
+    fieldName: string,
+    fieldConfig: ComposeFieldConfig<any, any>
+  ): DiscriminatorTypeComposer<TContext> {
+    super.setField(fieldName, fieldConfig);
+
+    for (const childTC of this.childTCs) {
+      childTC.setField(fieldName, fieldConfig);
```

Comment:
Here is an error.
You completely replace all fields in childTC, but it may have it own additional types. Code should be fixed in such way:

```diff
-      childTC.setField(fieldName, fieldConfig);
+      // childTC may have additional fields, so replace only Base fields
+      childTC.removeField(super.getFieldNames());
+      childTC.addFields(fields);
```

Also need tests for your methods.